src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2387,29 +2393,57 @@ static void
 ManageWorkerPool(WorkerPool *workerPool)
 {
 	DistributedExecution *execution = workerPool->distributedExecution;
-	int targetPoolSize = execution->targetPoolSize;
-	int initiatedConnectionCount = list_length(workerPool->sessionList);
-	int activeConnectionCount PG_USED_FOR_ASSERTS_ONLY =
-		workerPool->activeConnectionCount;
-	int idleConnectionCount PG_USED_FOR_ASSERTS_ONLY =
-		workerPool->idleConnectionCount;
-	int failedConnectionCount = workerPool->failedConnectionCount;
-	int readyTaskCount = workerPool->readyTaskCount;
-	int newConnectionCount = 0;
 
-	/* we should always have more (or equal) active connections than idle connections */
-	Assert(activeConnectionCount >= idleConnectionCount);
+	if (!CanOpenNewConnection(workerPool))
+	{
+		return;
+	}
 
-	/* we should always have more (or equal) initiated connections than active connections */
-	Assert(initiatedConnectionCount >= activeConnectionCount);
+	int newConnectionCount = CalculateNewConnectionCount(workerPool);
 
-	/* we should never have less than 0 connections ever */
-	Assert(activeConnectionCount >= 0 && idleConnectionCount >= 0);
+	if (newConnectionCount <= 0)
+	{
+		return;
+	}
+
+	OpenNewConnections(workerPool, newConnectionCount, execution->transactionProperties);
 
+	INSTR_TIME_SET_CURRENT(workerPool->lastConnectionOpenTime);
+	execution->rebuildWaitEventSet = true;
+}
+
+
+/*
+ * CanOpenNewConnection returns true if we can open new connections
+ * for the given workerPool.
+ */
+static bool
+CanOpenNewConnection(WorkerPool *workerPool)
+{
+	if (HasAnyConnectionFailure(workerPool))
+	{
+		return false;
+	}
+	if (ShouldWaitForSlowStart(workerPool))
```

Comment:
I'd rather inline these 2 checks in ManageWorkerPool so I don't need to wonder what "can open new connection" means, or how it is different from CalculateNewConnectionCount returning 0.